### PR TITLE
[Ecommerce] Return mock product in CartItem when actual product is not available

### LINKED ISSUE #9280: [Ecommerce] Error with deleted product in cart
# Bug Report

If you add an product to the cart and delete it in the backend it came to following error:
```
Call to a member function getOSName() on null
```

### Expected behavior
No error. Product should be removed.

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Resolves #9280

### CODE REVIEW

File: bundles/EcommerceFrameworkBundle/CartManager/SessionCart.php [hunk 1]
```diff
@@ -175,6 +177,14 @@ public function __wakeup()
 
         // set current cart
         foreach ($this->getItems() as $item) {
+
+            // if a product is deleted in backend, then remove item from SessionCart
+            $product = DataObject::getById($item->getProductId());
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/CartManager/SessionCart.php]: we should call `getProduct()` here ... then also PHPStan is happy ...

**Review by fashxp — CHANGES_REQUESTED** [review]

File: bundles/EcommerceFrameworkBundle/CartManager/SessionCart.php (hunk 1 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/CartManager/SessionCart.php]: done.

Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/CartManager/SessionCart.php]: Changed approach here as it impacts the performance coz when updating cart, it checks for product exists by calling `DataObject::getById()` which loads the product. 

so now we use `MockProduct` for CartItem in getProduct() calls when actual product doesn't exist.

File: bundles/EcommerceFrameworkBundle/Model/MockProduct.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+use Pimcore\Bundle\EcommerceFrameworkBundle\PriceSystem\PriceInterface;
+use Pimcore\Bundle\EcommerceFrameworkBundle\PriceSystem\PriceSystemInterface;
+use Pimcore\Model\DataObject\Concrete;
+
+/**
+ * Mock Product class which should be used as a product when actual product is not available in the System.
+ */
+class MockProduct extends Concrete implements ProductInterface, IndexableInterface, CheckoutableInterface
+{
+    public function getAvailabilitySystemName(): ?string
```
Comment (fashxp) [inline bundles/EcommerceFrameworkBundle/Model/MockProduct.php:31]: I would return null where ever possible. 
so also with `getAvailabilitySystemName`, `getPriceSystemImplementation`, `getAvailabilitySystemImplementation`, `getOSPriceInfo`, `getOSAvailabilityInfo`, `getPriceSystemName`

**Review by fashxp — CHANGES_REQUESTED** [review]

File: bundles/EcommerceFrameworkBundle/Model/MockProduct.php (hunk 2 above)
Comment (dvesh3) [reply bundles/EcommerceFrameworkBundle/Model/MockProduct.php:31]: It is not possible to return null for these methods due to incompatibility between CartItemInterface & CheckoutableInterface. I have created an issue for Pimcore 11 https://github.com/pimcore/pimcore/issues/9405

**Review by fashxp — APPROVED** [review]

**Comment by blankse** [comment]
@dvesh3 @fashxp AbstractCart::isValidCartItem() now is always true and can be removed?

**Comment by dvesh3** [comment]
@blankse valid point....but then let's just mark it deprecated coz removing it, is a BC break!

**Comment by blankse** [comment]
@dvesh3 BC break is also change the behavior. Should we check to MockProduct then?

**Comment by dvesh3** [comment]
@blankse 👍 added MockProduct check https://github.com/pimcore/pimcore/commit/9787cc1cabd1fe79e71ab2dfc8240251c135e16a, so the behavior is same as before. thanks for review 😉

**Pull request merged**